### PR TITLE
[EVM] Handle EVM errors

### PR DESCRIPTION
**Pull request by sideninja**
Closes: #5149 

This PR adds a type of EVMError that satisfies the UserError which is handled by Cadence at the same time embedding the CodedError.

The error output now looks like:
```
[Error Code: 1300] error caused by: 1 error occurred:
	* [Error Code: 1101] cadence runtime error: Execution failed:
 --> ff292f561c9cba05ef08abd74bc90cbaca1c217f737312d00b8fcf7ccafc405d:8:12
  |
8 | 				destroy acc.withdraw(balance: bal);
  | 				        ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: [Error Code: 1300] evm runtime error: insufficient total supply
  --> 8c5303eaa26202d6.EVM:89:12
   |
89 |             let vault <- InternalEVM.withdraw(
90 |                 from: self.addressBytes,
91 |                 amount: balance.flow
92 |             ) as! @FlowToken.Vault
   |             ^^^^^^^^^^^^^^^^^^^^^^


```

### CODE REVIEW

**Review by janezpodhostnik — APPROVED** [review]
Nice!

File: fvm/errors/errors.go [hunk 1]
```diff
@@ -213,3 +214,29 @@ func NewEventEncodingError(err error) CodedError {
...
+
+// EVMError needs to satisfy the user error interface
+// in order for Cadence to correctly handle the error
+var _ errors.UserError = &(EVMError{})
+
+type EVMError struct {
+	CodedError
+}
+
+func (e EVMError) IsUserError() {}
```
Comment (ramtinms) [inline fvm/errors/errors.go:226]: So is this the missing part ?

**Comment by ramtinms** [comment]
I think we need 3 tests
- one for when a fatal error is happening down there
- one for when a stateDB error is happening (non-fatal) 
- one for when a evm tx is failed

File: fvm/errors/errors.go (hunk 1 above)
Comment (sideninja) [reply fvm/errors/errors.go:226]: Basically, this makes it conform to the type of Cadence User Error which doesn't contain the stack and is handled gracefully. 
Now we could maybe even go further and adding an EVM error type in Cadence handling which I think could be even nicer way to go, but this solves it for now and we can add it to the todo if you think so?

**Comment by sideninja** [comment]
> I think we need 3 tests
> 
> * one for when a fatal error is happening down there
> * one for when a stateDB error is happening (non-fatal)
> * one for when a evm tx is failed

Ok, will add these tests as well.

**Comment by sideninja** [comment]
> happening down there

In the basement where all the EVM junk is 😆

File: fvm/fvm_test.go [hunk 2]
```diff
@@ -3013,4 +3013,102 @@ func TestEVM(t *testing.T) {
...
+	id flow.RegisterID,
+) (
+	flow.RegisterValue,
+	error,
+) {
+	if id.Key == "AddressAllocator" {
+		return nil, s.err
+	}
+
+	return s.snapshotTree.Get(id)
```
Comment (ramtinms) [inline fvm/fvm_test.go:3113]: hmm, I guess there should be a MockedSnapshot somewhere either created by mockery or implemented with func methods, If it doesn't exist we should create it in my view, that way we have more clear tests that are not dependent on a logic outside of the test.

Comment (ramtinms) [reply fvm/fvm_test.go:3113]: @janezpodhostnik do you know if we have one ?

Comment (sideninja) [reply fvm/fvm_test.go:3113]: I can replace with mockery yeah.

Comment (sideninja) [reply fvm/fvm_test.go:3113]: I've checked before for existing ones but only found, one in accounts that is just an implementation for tests here https://github.com/onflow/flow-go/blob/f17e073ae9d00e723a5d2e5fafe481a361030c5c/fvm/accounts_test.go#L22 and one in util https://github.com/onflow/flow-go/blob/0596ab06b1225f92b02f539234d292d2a02f619a/cmd/util/ledger/util/util.go#L67

So will generate one with mockery for now.

Comment (sideninja) [reply fvm/fvm_test.go]: @ramtinms fixed

**Review by turbolent — APPROVED** [review]
Nice!

**Review by ramtinms — APPROVED** [review]

**Pull request merged**